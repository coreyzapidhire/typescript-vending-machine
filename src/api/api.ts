import axios from "axios";
import { IItem } from "../types";

export const getProducts = async () => {
  return await axios
    .get("http://localhost:3001/products")
    .then((res) => res.data);
};

export const getProductById = async (id: number) => {
  return await axios
    .get(`http://localhost:3001/products/${id}`)
    .then((res) => res.data);
};

export const updateProduct = async (id: number, body: Partial<IItem>) => {
  return await axios
    .patch(`http://localhost:3001/products/${id}`, body)
    .then((res) => res.data);
};


