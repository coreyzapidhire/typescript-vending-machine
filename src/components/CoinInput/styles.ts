import styled from "styled-components";

export const AddCreditButton = styled.button`
  background-color: #4caf50;
  margin: 16px;
  border: none;
  color: white;
  padding: 15px 32px;
  text-align: center;
  text-decoration: none;
  display: inline-block;
  font-size: 32px;

  &:active {
    background-color: #4caf20;
  }
`;

export const InputWrapper = styled.div`
  display: "flex";
  flexdirection: "column";
`;

export const Input = styled.input`
  width: 20%;
  padding: 12px 20px;
  box-sizing: border-box;
  font-size: 32px;
`;

export const Title = styled.p``;
