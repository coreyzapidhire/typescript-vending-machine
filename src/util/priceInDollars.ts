import { IItem } from "../types";

export const amountInDollars = (priceInCents: number) => {
  const value = (priceInCents / 100).toLocaleString("en-US", {
    style: "currency",
    currency: "AUD",
  });
  return `${value}`;
};

export const canPurchase = (product: IItem, credit: number) =>
  credit >= product.priceInCents && product.stock > 0 ? true : false;
