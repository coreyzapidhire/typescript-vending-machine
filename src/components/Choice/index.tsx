import React, { useContext } from "react";
import { Text } from "../../common/styles";
import { RootContext } from "../../context/RootContext";
import { IItem } from "../../types";
import { amountInDollars } from "../../util/priceInDollars";
import { ChoiceWrapper } from "./styles";

export const Choice = ({ data, canBuy }: { data: IItem; canBuy: boolean }) => {
  const { purchaseItem } = useContext(RootContext);
  const { title, priceInCents, stock, id } = data;
  return (
    <>
      <ChoiceWrapper onClick={() => purchaseItem(id)} canBuy={canBuy}>
        <Text>{title}</Text>
        <Text>{amountInDollars(priceInCents)}</Text>
        <Text>{`${stock} in stock`}</Text>
      </ChoiceWrapper>
    </>
  );
};
