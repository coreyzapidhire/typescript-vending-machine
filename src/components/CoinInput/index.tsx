import React, { useContext, useState } from "react";
import { Container, Text } from "../../common/styles";
import { RootContext } from "../../context/RootContext";
import { EDenomination } from "../../types";
import { amountInDollars } from "../../util/priceInDollars";

import {
  AddCreditButton,
  Input,
  InputWrapper,
  RefundButton,
  Title,
} from "./styles";

export const CoinInput = () => {
  //local state
  const [currentEntry, setCurrentEntry] = useState("");
  const [changeReceived, setChangeReceived] = useState(false);
  //context
  const { credit, setCredit } = useContext(RootContext);

  //denominations
  const acceptedDenominations = [
    EDenomination.TenCents,
    EDenomination.TwentyCents,
    EDenomination.FiftyCents,
    EDenomination.OneDollar,
    EDenomination.TwoDollars,
  ];

  // handlers
  const addCreditHandler = () => {
    setChangeReceived(false);
    // check if credit is of correct denomination
    if (
      parseInt(currentEntry) !==
      acceptedDenominations.find((i) => i === parseInt(currentEntry))
    ) {
      // if incorrect denomination throw error
      // TODO add complete error display
      return alert(
        ` Must be ${acceptedDenominations.map((i: EDenomination) => `${i}`)}`
      );
    }
    // if all good then add credit
    setCredit(credit + parseInt(currentEntry));
  };

  const refundCreditHandler = () => {
    //refund credit
    setChangeReceived(true);
    setTimeout(() => setCredit(0), 3000);
  };

  return (
    <Container>
      <Container>
        {credit !== 0 && <Text>{`Credit: ${amountInDollars(credit)}`}</Text>}
      </Container>
      <InputWrapper>
        <Title>Enter Credit</Title>
        <Input
          title="Enter Amount"
          placeholder="Enter Amount"
          onChange={(e) => setCurrentEntry(e.target.value)}
          step="any"
        ></Input>
      </InputWrapper>
      <Container>
        <AddCreditButton onClick={addCreditHandler}>Add Credit</AddCreditButton>
        <RefundButton onClick={refundCreditHandler}>Refund</RefundButton>
      </Container>
      {changeReceived && (
        <Text> Received change: {amountInDollars(credit)}</Text>
      )}
    </Container>
  );
};
