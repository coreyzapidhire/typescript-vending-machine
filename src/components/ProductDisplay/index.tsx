import React, { useContext } from "react";
import { RootContext } from "../../context/RootContext";
import { IItem } from "../../types";
import { canPurchase } from "../../util/priceInDollars";
import { Choice } from "../Choice";
import { ProductDisplayContainer } from "./styles";

export const ProductDisplay = () => {
  const { products, credit } = useContext(RootContext);
  return (
    <ProductDisplayContainer>
      {products?.map((item: IItem, index) => {
        return (
          <Choice canBuy={canPurchase(item, credit)} key={index} data={item} />
        );
      })}
    </ProductDisplayContainer>
  );
};
