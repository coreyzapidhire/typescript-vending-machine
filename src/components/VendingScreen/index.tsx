import React from "react";
import { CoinInput } from "../CoinInput";
import { ProductDisplay } from "../ProductDisplay";

export const VendingScreen = () => {
  return (
    <>
      <CoinInput />
      <ProductDisplay />
    </>
  );
};
