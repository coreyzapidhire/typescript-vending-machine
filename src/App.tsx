import React from "react";
import "./App.css";
import { VendingScreen } from "./components/VendingScreen";
import { RootContextProvider } from "./context/RootContext";

function App() {
  return (
    <RootContextProvider>
      <div className="App">
        <header className="App-header">
          <VendingScreen />
        </header>
      </div>
    </RootContextProvider>
  );
}

export default App;
