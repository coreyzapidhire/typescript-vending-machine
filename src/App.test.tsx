import React from "react";
import { render, screen } from "@testing-library/react";
import App from "./App";
import renderer from 'react-test-renderer';
import { getProducts } from "./api/api";
import { amountInDollars } from "./util/priceInDollars";
import { CoinInput } from "./components/CoinInput";

test("test that data is returned from api", () => {
  const connectToServer = getProducts();

  expect(connectToServer).not.toBe(null);
});

test("correct credit is displayed", () => {
 expect(amountInDollars(200)).toBe('A$2.00')
 expect(amountInDollars(10)).toBe('A$0.10')

});

it('screen renders correctly', () => {
  const tree = renderer
    .create( <CoinInput />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
