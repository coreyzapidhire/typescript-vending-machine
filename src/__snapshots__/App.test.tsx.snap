// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`screen renders correctly 1`] = `
<div
  className="sc-bdfBwQ cmLqsX"
>
  <div
    className="sc-bdfBwQ cmLqsX"
  />
  <div
    className="sc-hKgILt eOexKO"
  >
    <p
      className="sc-jSgupP"
    >
      Enter Credit
    </p>
    <input
      className="sc-eCssSg ejiVVT"
      onChange={[Function]}
      placeholder="Enter Amount"
      step="any"
      title="Enter Amount"
    />
  </div>
  <button
    className="sc-dlfnbm bufZIA"
    onClick={[Function]}
  >
    Add Credit
  </button>
</div>
`;
